Describe actual Throwables when ThrowableAssert fails

### Summary
When some `Throwable`-related assertion fails the actual `Throwable` that failed the check is not described beyond what actually caused assertion failure (e.g. message mismatch). In non-trivial code when unexpected `Exception`s are thrown this makes debugging inconvenient.

Dumping more comprehensive (class, message, stack, perhaps even recursive info about causes) info about the actual `Throwable` could ease assertion failure investigation.

Relevant message factories (e.g. `ShouldHaveMessage`) already have access to actual `Throwable`.

It may be tempting to use `util.Throwables.ERROR_DESCRIPTION_EXTRACTOR` to describe throwables, but a more powerful mechanism would be appreciated. 
Providing precise information in case of assertion failure doesn't seem to be overkill (unlike as in #864)

### Error message to improve

- [x] ShouldHaveCause #2872
- [ ] ShouldHaveCauseExactlyInstance
- [ ] ShouldHaveCauseInstance
- [ ] ShouldHaveCauseReference - @weiyilei
- [ ] ShouldHaveNoCause
- [ ] ShouldHaveNoSuppressedExceptions
- [x] ShouldHaveRootCauseExactlyInstance #2910
- [x] ShouldHaveRootCauseInstance #2910
- [ ] ShouldHaveSuppressedException
- [ ] ShouldNotContainCharSequence 
- [ ] ShouldStartWith
- [ ] ShouldEndWith

If you are interested in contributing to this issue, select one error message, it will be assigned to you.

Example of my ideal message:
```java
Expecting Throwable message:
<"expected exception message">
but was:
<"some other unexpected exception">

Throwable that failed the check:
Exception in thread "main" java.lang.RuntimeException: some other unexpected exception
	at Test.functionThatIsSupposedToThrowExpectedException(Test.java:119)
	at Test.main(Test.java:69)
Caused by: java.lang.RuntimeException: unexpected exception in recurseToGenerateSomeStack
	at Test.recurseToGenerateSomeStack(Test.java:133)
	at Test.recurseToGenerateSomeStack(Test.java:138)
	at Test.recurseToGenerateSomeStack(Test.java:138)
	at Test.recurseToGenerateSomeStack(Test.java:138)
	at Test.recurseToGenerateSomeStack(Test.java:138)
	at Test.recurseToGenerateSomeStack(Test.java:138)
	at Test.recurseToGenerateSomeStack(Test.java:138)
	at Test.recurseToGenerateSomeStack(Test.java:138)
	at Test.recurseToGenerateSomeStack(Test.java:138)
	at Test.recurseToGenerateSomeStack(Test.java:138)
	at Test.recurseToGenerateSomeStack(Test.java:138)
	at Test.recurseToGenerateSomeStack(Test.java:138)
	at Test.functionThatIsSupposedToThrowExpectedException(Test.java:115)
	... 1 more
Caused by: java.lang.RuntimeException: the root cause of test failure
	at Test.f(Test.java:169)
	at Test.e(Test.java:164)
	at Test.d(Test.java:159)
	at Test.c(Test.java:154)
	at Test.b(Test.java:149)
	at Test.a(Test.java:144)
	at Test.recurseToGenerateSomeStack(Test.java:129)
	... 13 more
```

Such message allows to immediately learn the real cause of test failure.

--- a/assertj-core/src/main/java/org/assertj/core/error/ShouldNotContainCharSequence.java
+++ b/assertj-core/src/main/java/org/assertj/core/error/ShouldNotContainCharSequence.java
@@ -69,7 +69,7 @@ public class ShouldNotContainCharSequence extends BasicErrorMessageFactory {
                     "  %s%n" +
                     "not to contain:%n" +
                     "  %s%n" +
-                    "but found:%n" +
+                    "but did:%n" +
                     "%n" +
                     "Throwable that failed the check:%n" +
                     "%n" + escapePercent(getStackTrace(actual)); // to avoid AssertJ default String formatting

--- a/assertj-core/src/test/java/org/assertj/core/error/ShouldNotContainThrowable_create_Test.java
+++ b/assertj-core/src/test/java/org/assertj/core/error/ShouldNotContainThrowable_create_Test.java
@@ -33,16 +33,16 @@ class ShouldNotContainThrowable_create_Test {
   @Test
   void should_create_error_message_with_escaping_percent() {
     // GIVEN
-    RuntimeException actual = new RuntimeException("You know nothing %");
+    RuntimeException actual = new RuntimeException("You know nothing % Jon Snow");
     // WHEN
-    String errorMessage = shouldNotContain(actual, "You know nothing % Jon Snow").create(new TestDescription("TEST"));
+    String errorMessage = shouldNotContain(actual, "You know nothing %").create(new TestDescription("TEST"));
     // THEN
     then(errorMessage).isEqualTo("[TEST] %n" +
                                  "Expecting throwable message:%n" +
-                                 "  \"You know nothing %%\"%n" +
-                                 "not to contain:%n" +
                                  "  \"You know nothing %% Jon Snow\"%n" +
-                                 "but found:%n" +
+                                 "not to contain:%n" +
+                                 "  \"You know nothing %%\"%n" +
+                                 "but did:%n" +
                                  "%n" +
                                  "Throwable that failed the check:%n" +
                                  "%n%s", getStackTrace(actual));
@@ -51,19 +51,19 @@ class ShouldNotContainThrowable_create_Test {
   @Test
   void should_create_error_message_with_several_values_not_found() {
     // GIVEN
-    RuntimeException actual = new RuntimeException("You know nothing");
-    String[] sequence = array("You", "know", "nothing", "Jon", "Snow");
-    Set<String> notFound = newSet("Jon", "Snow");
+    RuntimeException actual = new RuntimeException("You know nothing Jon Snow");
+    String[] sequence = array("You", "know", "nothing");
+    Set<String> found = newSet("You", "know", "nothing");
     // WHEN
-    String errorMessage = shouldNotContain(actual, sequence, notFound).create(new TestDescription("TEST"));
+    String errorMessage = shouldNotContain(actual, sequence, found).create(new TestDescription("TEST"));
     // THEN
     then(errorMessage).isEqualTo("[TEST] %n" +
                                  "Expecting throwable message:%n" +
-                                 "  \"You know nothing\"%n" +
+                                 "  \"You know nothing Jon Snow\"%n" +
                                  "not to contain:%n" +
-                                 "  [\"You\", \"know\", \"nothing\", \"Jon\", \"Snow\"]%n" +
+                                 "  [\"You\", \"know\", \"nothing\"]%n" +
                                  "but found:%n" +
-                                 "  [\"Jon\", \"Snow\"]%n" +
+                                 "  [\"You\", \"know\", \"nothing\"]%n" +
                                  "%n" +
                                  "Throwable that failed the check:%n" +
                                  "%n%s", getStackTrace(actual));

